src/lib.rs

```diff
@@ -2,9 +2,40 @@
 #![feature(exact_chunks)]
 
 extern crate test;
+extern crate faster;
+
+use faster::*;
 
 const RGB_Y: [u32; 4] = [19595, 38470, 7471, 0];
 
+pub fn bgrx_to_gray_chunks_no_asserts_faster(in_data: &[u8], out_data: &mut [u8], in_stride: usize, out_stride: usize, width: usize) {
+    // Sane, but slowed down by faster's current striping implementation.
+    in_data.stripe_four(tuplify!(4, u8s(0))).zip().simd_map(|(r, g, b, _)| {
+        let (r16a, r16b) = r.upcast();
+        let (r32a, r32b) = r16a.upcast();
+        let (r32c, r32d) = r16b.upcast();
```

Comment:
So `r` is basically a `[u8; 4]`, `r16a` a `[u16; 2]`, etc? The approach faster takes here seems like the reverse of what GStreamer's [orc](https://cgit.freedesktop.org/gstreamer/orc) is doing, where you would specify that you read `u32` from the source array and then have a feature to interpret the `u32` as `[u8; 4]` or `[u16; 2]` and then can operate on those. So you can do the same operation on the whole `[u8; 4]` in a single step, instead of working on the individual components, which can probably allow for some further optimizations? As the individual steps as in your PR for each of the 4 pixels could in theory be different. Basically what you could do is to map all this to `(rgba * coefficients) / 65536` (where `*` and `/` are component-wise) and then downcasting the whole thing

See for example [this](https://cgit.freedesktop.org/gstreamer/gst-plugins-good/tree/gst/videomixer/videomixerorc.orc?h=84ec78bd862192b87ba8b35732265f1511a1427f#n32), the `x4` prefix interprets things as "components". It's basically blending the `d` and `s` ARGB pixels and multiplies `s` with an alpha value.

Just FYI for how these things were designed elsewhere, in case it gives you some ideas or insights for `faster` :) In case of orc this processing model mapped well to e.g. SSE and NEON.

Reply:
Yeah, I usually try to use strided loads with faster regardless of the data model because it's theoretically more portable (although I've never seen a vector width or screen resolution which is indivisible by 4).

Reply:
I would be surprised to see vector sizes that are not a power of two, but who knows :)

For the frame resolution, that can happen indeed. What ORC is doing in such cases is that it first processes everything it can with SIMD instructions, and then does the remainder one-by-one.